portfolyo/tools/freq.py

```diff
@@ -7,15 +7,90 @@
+    pd.tseries.offsets.MonthBegin,
+    pd.tseries.offsets.Day,
+    pd.tseries.offsets.Hour,
+    pd.tseries.offsets.Minute,
+]
+TO_OFFSET = pd.tseries.frequencies.to_offset
+SHORTEST_TO_LONGEST = [
+    type(TO_OFFSET(freq)) for freq in ["15T", "H", "D", "MS", "QS", "AS"]
+]
+
+quarter_matrix = [
+    ["QS", "QS-APR", "QS-JUL", "QS-OCT"],
+    ["QS-FEB", "QS-MAY", "QS-AUG", "QS-NOV"],
+    ["QS-MAR", "QS-JUN", "QS-SEP", "QS-DEC"],
+]
+
+
+def assert_freq_valid(freq: str) -> None:
+    """
+    Validate if the given frequency string is allowed based on pandas offset objects.
 
+    Parameters:
+    freq (str): A string representing a frequency alias (e.g., "AS", "QS", "MS").
 
-def up_or_down(
-    freq_source: str, freq_target: str, common_ts: pd.Timestamp = None
-) -> int:
+    Raises:
+    ValueError: If the frequency is not allowed.
+    """
+
+    freq_offset = pd.tseries.frequencies.to_offset(freq)
+    mro_class = freq_offset.__class__.__mro__[0]
+
+    # Check if the MRO is in the list of allowed MROs
+    # have to make sure it's only the first class on the list
+    if mro_class not in ALLOWED_CLASSES:
+        raise ValueError(f"The passed frequency '{freq}' is not allowed.")
+
+    # Define restricted classes that should have n == 1
+    restricted_classes = {
+        pd.tseries.offsets.MonthBegin: 1,
+        pd.tseries.offsets.Day: 1,
+        pd.tseries.offsets.Hour: 1,
+        pd.tseries.offsets.Minute: 15,
+    }
+    allowed_n = restricted_classes.get(type(freq_offset))
+    if allowed_n is not None:  # case where freq is not in restricted class
+        # Check if freq_offset.n is not None and if it doesn't match allowed_n
+        if freq_offset.n is None or freq_offset.n != allowed_n:
+            raise ValueError(f"The passed frequency {freq} is not allowed.")
+
+
+def assert_freq_sufficiently_long(freq, freq_ref, strict: bool = False) -> None:
+    """
+    Compares ``freq`` and ``freq_ref``, raising an AssertionError if ``freq`` is not long enough.
+
+    Parameters
+    ----------
+    freq_source, freq_ref : frequencies to compare.
+    strict : bool, optional (default: False)
+        - If ``strict`` is True, ``freq`` must be strictly longer than ``freq_long``.
+        - If False, it may be equally long.
+
+    """
+    # freq should start from the beginning of the year
+    index_freq = SHORTEST_TO_LONGEST.index(type(TO_OFFSET(freq)))
+    index_ref = SHORTEST_TO_LONGEST.index(type(TO_OFFSET(freq_ref)))
+    if strict is True:
+        if not (index_freq > index_ref):
+            raise AssertionError(
+                f"The passed frequency is not sufficiently long; passed {freq}, but should be {freq_ref} or longer."
+            )
+    else:
+        if not (index_freq >= index_ref):
+            raise AssertionError(
+                f"The passed frequency is not sufficiently long; passed {freq}, but should be {freq_ref} or longer."
+            )
+
+
+def up_or_down2(freq_source: str, freq_target: str) -> int:
```

Comment:
Since we no longer have `up_or_down`, maybe we can rename this back to `up_or_down`? And move it down in the file, so `assert_sufficiently_long` is next to `assert_sufficiently_short` and `assert_equally_long`.

Reply:
done